cosi/cosi.go

```diff
@@ -0,0 +1,502 @@
+/*
+Package cosi is the Collective Signing implementation according to the paper of
+Bryan Ford: http://arxiv.org/pdf/1503.08768v1.pdf .
+
+Stages of CoSi
+
+The CoSi-protocol has 4 stages:
+
+1. Announcement: The leader multicasts an announcement
+of the start of this round down through the spanning tree,
+optionally including the statement S to be signed.
+
+2. Commitment: Each node i picks a random secret vi and
+computes its individual commit Vi = Gvi . In a bottom-up
+process, each node i waits for an aggregate commit Vˆj from
+each immediate child j, if any. Node i then computes its
+own aggregate commit Vˆi = Vi \prod{j ∈ Cj}{Vˆj}, where Ci is the
+set of i’s immediate children. Finally, i passes Vi up to its
+parent, unless i is the leader (node 0).
+
+3. Challenge: The leader computes a collective challenge
+c = H( Aggregate Commit ∥ Aggregate Public key || Message ),
+then multicasts c down through the tree, along
+with the statement S to be signed if it was not already
+announced in phase 1.
+
+4. Response: In a final bottom-up phase, each node i waits
+to receive a partial aggregate response rˆj from each of
+its immediate children j ∈ Ci. Node i now computes its
+individual response ri = vi + cxi, and its partial aggregate
+response rˆi = ri + \sum{j ∈ Cj}{rˆj} . Node i finally passes rˆi
+up to its parent, unless i is the root.
+*/
+package cosi
+
+import (
+	cryptoRand "crypto/rand"
+	"crypto/sha512"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/dedis/crypto/abstract"
+	//own "github.com/nikkolasg/learning/crypto/util"
+)
+
+// Cosi is the struct that implements the basic cosi.
+type Cosi struct {
```

Comment:
Also, please make the comment more descriptive.  Yes, it's obviously a struct, but what's "the basic cosi"?  The struct is one-per-what?  One per protocol instance?  One per protocol run?  What?  Also summarize briefly how to use it: the way(s) to create one, the life-cycle of the struct, etc.  If the caller is supposed to invoke methods in sequence for the four stages mentioned in the module godoc, then no need to duplicate all that info, but at least list the 4 specific function names the caller is expected to call for each of the 4 stages, etc.